File: StreamingBrowserElectron/src/startphasen.js
"use strict";

/**
 * Die Ladephasen beim Start einer Folge.
 *
 * <h2>Warum es dieses Modul gibt</h2>
 *
 * Zwischen "Weiterschauen" gedrueckt und "das Video laeuft im Vollbild" liegen
 * mehrere Schritte, die keiner Uhr folgen: die Folgenseite muss geladen sein,
 * der Hoster muss seinen Player bauen, der Player muss seine Quelle holen, der
 * gespeicherte Stand muss gesetzt werden, und erst danach ergibt Vollbild
 * ueberhaupt einen Sinn. Waehrenddessen liegt auf beiden Geraeten ein Vorhang
 * davor - sonst sieht man den Seitenaufbau, die Ueberlagerung des Hosters und
 * den Sprung ins Vollbild.
 *
 * Ein Ladebalken, der dabei einfach hochzaehlt, waere gelogen: er saehe bei
 * einer Leitung, die dreissig Sekunden braucht, genauso aus wie bei einer, die
 * drei braucht. Deshalb zaehlt hier nicht die Zeit, sondern die *Phase*. Jede
 * Phase entspricht einem Schritt, den die Startkette wirklich hinter sich
 * gebracht hat; der Balken springt weiter, wenn ein Schritt gemeldet wird, und
 * sonst gar nicht.
 *
 * <h2>Warum es geteilt ist</h2>
 *
 * Die Kette gibt es zweimal - main.js am Rechner, MainActivity und Mitschauen
 * auf Android -, aber sie hat dieselben Schritte, und was der Zuschauer dabei
 * liest, soll auf beiden Geraeten dasselbe sein. Namen, Reihenfolge,
 * Beschriftungen und Fristen stehen darum hier und nur hier. Android holt die
 * Tabelle beim Start einmal ueber den Kern ab (siehe Startphasen.java), der
 * Rechner ruft die Funktionen unmittelbar auf.
 *
 * Reine Rechnung, kein DOM, kein Node: das Modul laeuft im Kern-WebView
 * genauso wie im Hauptprozess, und es laesst sich ohne Geraet pruefen
 * (tests/startphasentest.js).
 */

/**
 * Die Schritte, in ihrer Reihenfolge.
 *
 * `fristMs` ist die Geduld fuer genau diesen Schritt - nicht fuer den ganzen
 * Start. Sie sind absichtlich unterschiedlich: eine Folgenseite darf lange
 * laden (gemessen wurden auf einer gedrosselten Leitung ueber 150 Sekunden bis
 * zum letzten Werbebild, waehrend der Teil, auf den es ankommt, laengst
 * dastand), ein Vollbild dagegen ist entweder sofort da oder gar nicht.
 *
 * `anteil` ist der Stand des Balkens, *wenn diese Phase beginnt*. Bewusst als
 * Tabelle und nicht als Nummer durch Laenge: die Schritte dauern
 * unterschiedlich lang, und ein Balken, der beim Warten auf den Hoster in der
 * Mitte steht, sagt mehr als einer, der in gleichen Fuenfteln springt.
 */
const PHASEN = [
  { name: "seite", text: "Folge wird geöffnet", anteil: 0.08, fristMs: 90000 },
  { name: "hoster", text: "Player wird geladen", anteil: 0.32, fristMs: 60000 },
  { name: "spieler", text: "Video wird vorbereitet", anteil: 0.58, fristMs: 45000 },
  { name: "stelle", text: "Zur gespeicherten Stelle", anteil: 0.78, fristMs: 25000 },
  { name: "vollbild", text: "Vollbild wird gesetzt", anteil: 0.90, fristMs: 15000 },
  { name: "laeuft", text: "Wiedergabe läuft", anteil: 1, fristMs: 0 }
];

/** Die Phase, in der ein Start anfaengt. */
const ERSTE = PHASEN[0].name;

/** Die Phase, nach der es nichts mehr zu tun gibt. */
const LETZTE = PHASEN[PHASEN.length - 1].name;

/**
 * Der Deckel ueber allem.
 *
 * Die Einzelfristen koennen sich addieren - jede fuer sich plausibel, in Summe
 * aber laenger, als irgendjemand vor einem Ladebildschirm sitzen will. Nach
 * dieser Zeit ist Schluss, egal in welchem Schritt es haengt.
 */
const GESAMT_FRIST_MS = 150000;

function phaseAus(name) {
  return PHASEN.find((phase) => phase.name === String(name || "")) || null;
}

function nummer(name) {
  return PHASEN.findIndex((phase) => phase.name === String(name || ""));
}

/**
 * Welche Schritte dieser Start durchlaeuft.
 *
 * Ohne gespeicherten Stand gibt es nichts zu springen - dann faellt "stelle"
 * heraus, statt als Schritt dazustehen, den niemand je meldet.
 *
 * @param {{stelle?: number}} optionen
 */
function phasen(optionen = {}) {
  const stelle = Number(optionen.stelle) || 0;
  return PHASEN
    .filter((phase) => phase.name !== "stelle" || stelle > 0)
    .map((phase) => ({ ...phase }));
}

/** Was in dieser Phase dasteht. Leer, wenn der Name keiner ist. */
function beschriftung(name) {
  const phase = phaseAus(name);
  return phase ? phase.text : "";
}

/** Wie voll der Balken in dieser Phase ist, 0 bis 1. */
function anteil(name) {
  const phase = phaseAus(name);
  return phase ? phase.anteil : 0;
}

/**
 * Einen Start anlegen.
 *
 * @param {{titel?: string, stelle?: number, jetzt?: number}} angaben
 */
function starten(angaben = {}) {
  const jetzt = Number(angaben.jetzt) || 0;
  return {
    titel: String(angaben.titel || ""),
    stelle: Math.max(0, Number(angaben.stelle) || 0),
    phase: ERSTE,
    begonnen: jetzt,
    seit: jetzt,
    fertig: false,
    fehler: ""
  };
}

/**
 * Einen Schritt melden.
 *
 * Rueckwaerts geht es nicht: die Kette meldet manche Schritte mehrfach (jeder
 * Takt des Autostarts sieht denselben Zustand wieder), und ein Balken, der
 * dabei zurueckspringt, sieht aus wie ein Fehler. Eine Meldung, die hinter dem
 * Stand liegt, wird darum still verworfen.
 *
 * @return {{zustand, geaendert: boolean, phase: string, text: string, anteil: number, fertig: boolean}}
 */
function melden(zustand, name, jetzt = 0) {
  const antwort = (geaendert) => ({
    zustand,
    geaendert,
    phase: zustand ? zustand.phase : "",
    text: beschriftung(zustand ? zustand.phase : ""),
    anteil: anteil(zustand ? zustand.phase : ""),
    fertig: Boolean(zustand && zustand.fertig)
  });
  if (!zustand || zustand.fehler) return antwort(false);
  const ziel = nummer(name);
  if (ziel < 0) return antwort(false);
  if (ziel <= nummer(zustand.phase)) return antwort(false);
  zustand.phase = PHASEN[ziel].name;
  zustand.seit = Number(jetzt) || 0;
  if (zustand.phase === LETZTE) zustand.fertig = true;
  return antwort(true);
}

/**
 * Steht der Start noch, oder ist er haengengeblieben?
 *
 * Zwei Fristen, und beide muessen sein: die des Schrittes faengt den Fall, in
 * dem ein einzelner Schritt nie antwortet, der Gesamtdeckel den, in dem jeder
 * Schritt gerade noch rechtzeitig kommt und es trotzdem ewig dauert.
 *
 * @return {{tun: "warten"|"fehler", grund: string, text: string}}
 */
function pruefen(zustand, jetzt = 0) {
  if (!zustand) return { tun: "fehler", grund: "kein start", text: fehlertext("kein start") };
  if (zustand.fertig) return { tun: "warten", grund: "fertig", text: "" };
  if (zustand.fehler) return { tun: "fehler", grund: zustand.fehler, text: fehlertext(zustand.fehler) };
  const zeit = Number(jetzt) || 0;
  // Ohne Kurzschluss auf `begonnen`: eine Uhr, die bei null anfaengt, ist eine
  // gueltige Uhr. Mit `zustand.begonnen &&` davor war der Deckel genau dann
  // abgeschaltet, wenn der Start im ersten Zeittakt begonnen hatte - und in
  // jeder Pruefung, die bei null zu zaehlen anfaengt.
  if (zeit - zustand.begonnen >= GESAMT_FRIST_MS) {
    return { tun: "fehler", grund: "gesamt", text: fehlertext("gesamt") };
  }
  const phase = phaseAus(zustand.phase);
  if (phase && phase.fristMs > 0 && zeit - zustand.seit >= phase.fristMs) {
    return { tun: "fehler", grund: zustand.phase, text: fehlertext(zustand.phase) };
  }
  return { tun: "warten", grund: zustand.phase, text: beschriftung(zustand.phase) };
}

/**
 * Was dem Zuschauer gesagt wird, wenn es nicht weitergeht.
 *
 * Je Schritt eine eigene Antwort. "Zeitueberschreitung" waere richtig und
 * nutzlos - wer liest, dass der Hoster keinen Player geliefert hat, weiss,
 * dass ein zweiter Versuch etwas bringen kann, und wer liest, dass die Seite
 * nicht laedt, sieht zuerst nach seiner Verbindung.
 */
function fehlertext(grund) {
  switch (String(grund || "")) {
    case "seite":
      return "Die Folgenseite lädt nicht. Prüfe deine Internetverbindung.";
    case "hoster":
      return "Der Hoster hat keinen Player geliefert.";
    case "spieler":
      return "Der Player hat kein Video geladen.";
    case "stelle":
      return "Der Player ist nicht zur gespeicherten Stelle gesprungen.";
    case "vollbild":
      return "Das Video läuft, aber das Vollbild kam nicht zustande.";
    case "gesamt":
      return "Der Start dauert zu lange.";
    case "abgebrochen":
      return "Der Start wurde abgebrochen.";
    default:
      return "Die Folge konnte nicht gestartet werden.";
  }
}

/**
 * Die ganze Tabelle als schlichte Angaben - fuer Aufrufer ohne JavaScript.
 *
 * Android liest sie beim Start einmal aus und zeichnet danach ohne Rueckfrage.
 * Ein Weg ueber den Kern je Balkenschritt waere ein Rundlauf fuer eine
 * Division, und der Vorhang muss sofort dastehen.
 */
function modell() {
  return {
    phasen: PHASEN.map((phase) => ({ ...phase })),
    erste: ERSTE,
    letzte: LETZTE,
    gesamtFristMs: GESAMT_FRIST_MS,
    fehlertexte: PHASEN
      .map((phase) => phase.name)
      .concat(["gesamt", "abgebrochen", ""])
      .reduce((sammlung, name) => {
        sammlung[name] = fehlertext(name);
        return sammlung;
      }, {})
  };
}

module.exports = {
  PHASEN,
  ERSTE,
  LETZTE,
  GESAMT_FRIST_MS,
  phasen,
  beschriftung,
  anteil,
  starten,
  melden,
  pruefen,
  fehlertext,
  modell
};

File: android/app/src/main/java/local/elflix/android/Startphasen.java
package local.elflix.android;

import android.util.Log;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Die Ladephasen beim Start einer Folge - so, wie der Rechner sie kennt.
 *
 * <h2>Was hier steht und was nicht</h2>
 *
 * <p>Hier steht <em>keine</em> Tabelle. Namen, Reihenfolge, Beschriftungen,
 * Fristen und Fehlertexte kommen aus {@code src/startphasen.js}, demselben
 * Modul, das der Rechner benutzt; diese Klasse holt sie beim Start einmal ueber
 * den {@link Kern} ab und haelt sie. Wer eine Beschriftung aendern will,
 * aendert sie dort, und beide Geraete sagen danach dasselbe.
 *
 * <p>Gerechnet wird danach hier - und zwar genau zweierlei: ob eine gemeldete
 * Phase weiter ist als die stehende (ein Vergleich zweier Nummern) und ob eine
 * Frist abgelaufen ist (eine Subtraktion). Ein Weg ueber die Bruecke waere
 * dafuer ein Rundlauf pro Balkenschritt, und der Vorhang muss sofort
 * dastehen. Dieselbe Abwaegung wie bei {@code Folgen.prozent}; damit sie nicht
 * auseinanderlaufen kann, ist die Tabelle die geteilte und nur die Mechanik
 * hiesig.
 *
 * <h2>Ohne Kern kein Vorhang</h2>
 *
 * <p>{@link #istBereit()} bleibt {@code false}, solange die Tabelle nicht da
 * ist. Das ist kein Notfall, sondern der ehrliche Zustand: ohne Kern laeuft
 * auch die Startkette selbst nicht (siehe {@code Mitschauen.oertlichenStartAnfordern}),
 * es gibt also nichts zu begleiten. Der Vorhang bleibt dann weg, und der Start
 * verhaelt sich wie vor dieser Aenderung.
 */
final class Startphasen {
    private static final String TAG = CrashReporter.TAG;

    /** Ein Schritt, wie ihn das geteilte Modul beschreibt. */
    static final class Phase {
        final String name;
        final String text;
        /** Wie voll der Balken ist, wenn dieser Schritt beginnt - 0 bis 1. */
        final double anteil;
        /** Die Geduld fuer genau diesen Schritt. 0 heisst: keine Frist. */
        final long fristMs;

        Phase(String name, String text, double anteil, long fristMs) {
            this.name = name;
            this.text = text;
            this.anteil = anteil;
            this.fristMs = fristMs;
        }
    }

    private final Kern kern;
    private final List<Phase> phasen = new ArrayList<>();
    private final Map<String, String> fehlertexte = new HashMap<>();
    private long gesamtFristMs;

    Startphasen(Kern kern) {
        this.kern = kern;
    }

    /**
     * Die Tabelle holen. Einmal, beim Start, sobald der Kern steht.
     *
     * <p>Fehlschlaege sind still bis auf die Protokollzeile: sie bedeuten nur,
     * dass es keinen Vorhang gibt, und ein Hinweis darueber waere eine Meldung
     * ueber etwas, das der Zuschauer nie angefordert hat.
     */
    void vorbereiten() {
        if (kern == null) return;
        kern.wennBereit(() -> kern.rufe("startphasen.modell", (wert, fehler) -> {
            if (fehler != null || wert == null) {
                Log.w(TAG, "Startphasen nicht geladen: " + fehler);
                return;
            }
            try {
                uebernehmen(new JSONObject(wert));
            } catch (Exception ausnahme) {
                Log.w(TAG, "Startphasen unlesbar", ausnahme);
            }
        }));
    }

    /** Sichtbar fuer die Pruefung: dieselbe Uebernahme, ohne Kern. */
    void uebernehmen(JSONObject modell) {
        if (modell == null) return;
        List<Phase> gelesen = new ArrayList<>();
        JSONArray liste = modell.optJSONArray("phasen");
        for (int i = 0; liste != null && i < liste.length(); i += 1) {
            JSONObject eintrag = liste.optJSONObject(i);
            if (eintrag == null) continue;
            String name = eintrag.optString("name", "");
            if (name.isEmpty()) continue;
            gelesen.add(new Phase(
                name,
                eintrag.optString("text", ""),
                eintrag.optDouble("anteil", 0),
                eintrag.optLong("fristMs", 0)));
        }
        if (gelesen.isEmpty()) return;
        phasen.clear();
        phasen.addAll(gelesen);
        gesamtFristMs = modell.optLong("gesamtFristMs", 0);
        fehlertexte.clear();
        JSONObject texte = modell.optJSONObject("fehlertexte");
        if (texte != null) {
            for (java.util.Iterator<String> it = texte.keys(); it.hasNext(); ) {
                String schluessel = it.next();
                fehlertexte.put(schluessel, texte.optString(schluessel, ""));
            }
        }
        Log.i(TAG, "Startphasen geladen: " + phasen.size() + " Schritte, Deckel "
            + gesamtFristMs + " ms");
    }

    boolean istBereit() {
        return !phasen.isEmpty();
    }

    /** Der Name des ersten Schrittes - damit faengt jeder Start an. */
    String erste() {
        return phasen.isEmpty() ? "" : phasen.get(0).name;
    }

    /** Der Name des letzten Schrittes. Ist er erreicht, ist der Start zu Ende. */
    String letzte() {
        return phasen.isEmpty() ? "" : phasen.get(phasen.size() - 1).name;
    }

    long gesamtFristMs() {
        return gesamtFristMs;
    }

    /** Die Nummer eines Schrittes, oder -1. Nur damit sich Schritte vergleichen lassen. */
    int nummer(String name) {
        for (int i = 0; i < phasen.size(); i += 1) {
            if (phasen.get(i).name.equals(name)) return i;
        }
        return -1;
    }

    Phase phase(String name) {
        int nummer = nummer(name);
        return nummer < 0 ? null : phasen.get(nummer);
    }

    /** Was in diesem Schritt dasteht. */
    String beschriftung(String name) {
        Phase phase = phase(name);
        return phase == null ? "" : phase.text;
    }

    /** Wie voll der Balken in diesem Schritt ist, 0 bis 1. */
    double anteil(String name) {
        Phase phase = phase(name);
        return phase == null ? 0 : phase.anteil;
    }

    /** Was dem Zuschauer gesagt wird, wenn es an dieser Stelle nicht weitergeht. */
    String fehlertext(String grund) {
        String text = fehlertexte.get(grund == null ? "" : grund);
        if (text != null && !text.isEmpty()) return text;
        String allgemein = fehlertexte.get("");
        return allgemein == null ? "" : allgemein;
    }
}
